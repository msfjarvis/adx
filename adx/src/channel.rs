use crate::version::Version;
use clap::ValueEnum;
use clap::builder::PossibleValue;
use semver::Prerelease;
use std::cmp::{Eq, PartialEq, PartialOrd};
use std::convert::TryFrom;
use std::fmt::Debug;
use std::str::FromStr;
use thiserror::Error;

/// Release channels for androidx packages
/// Since we're deriving `PartialOrd` automatically, the order
/// of these fields is crucial. Sort by stability, not alphabetical
/// order.
#[derive(Debug, Eq, PartialEq, PartialOrd, Clone, Copy)]
pub(crate) enum Channel {
    Dev,
    Alpha,
    Beta,
    Rc,
    Stable,
}

impl ValueEnum for Channel {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::Dev, Self::Alpha, Self::Beta, Self::Rc, Self::Stable]
    }

    fn to_possible_value(&self) -> Option<PossibleValue> {
        match self {
            Channel::Dev => Some(PossibleValue::new("dev").alias("d")),
            Channel::Alpha => Some(PossibleValue::new("alpha").alias("a")),
            Channel::Beta => Some(PossibleValue::new("beta").alias("b")),
            Channel::Stable => Some(PossibleValue::new("stable").alias("s")),
            Channel::Rc => Some(PossibleValue::new("rc").alias("r")),
        }
    }
}

#[derive(Debug, Error)]
pub(crate) enum ChannelError {
    #[error("no match found")]
    NoMatchFound,
    #[error("failed to determine channel for {0}")]
    FailedToParseVersion(Version),
}

impl FromStr for Channel {
    type Err = ChannelError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "alpha" | "a" => Ok(Channel::Alpha),
            "beta" | "b" => Ok(Channel::Beta),
            "dev" | "d" => Ok(Channel::Dev),
            "rc" | "r" => Ok(Channel::Rc),
            "stable" | "s" => Ok(Channel::Stable),
            _ => Err(ChannelError::NoMatchFound),
        }
    }
}

impl TryFrom<&Version> for Channel {
    type Error = ChannelError;

    fn try_from(value: &Version) -> Result<Self, Self::Error> {
        match value {
            Version::SemVer(version) => {
                if version.pre == Prerelease::EMPTY {
                    Ok(Channel::Stable)
                } else {
                    let pre_str = version.pre.to_string();
                    if pre_str.starts_with("alpha") {
                        Ok(Channel::Alpha)
                    } else if pre_str.starts_with("beta") {
                        Ok(Channel::Beta)
                    } else if pre_str.starts_with("dev") {
                        Ok(Channel::Dev)
                    } else if pre_str.starts_with("rc") {
                        Ok(Channel::Rc)
                    } else {
                        Err(ChannelError::FailedToParseVersion(value.clone()))
                    }
                }
            }
            Version::CalVer(_) => Ok(Channel::Stable),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Channel;
    use crate::version::Version;
    use semver::{BuildMetadata, Prerelease, Version as Semver};
    use std::convert::TryFrom;

    #[test]
    fn alpha_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("alpha01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Alpha, channel.unwrap());
    }

    #[test]
    fn beta_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("beta01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Beta, channel.unwrap());
    }

    #[test]
    fn dev_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("dev01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Dev, channel.unwrap());
    }

    #[test]
    fn rc_version() {
        let v = Version::SemVer(Semver {
            major: 1,
            minor: 1,
            patch: 0,
            pre: Prerelease::new("rc01").unwrap(),
            build: BuildMetadata::EMPTY,
        });
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Rc, channel.unwrap());
    }

    #[test]
    fn stable_version() {
        let v = Version::SemVer(Semver::new(1, 1, 1));
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Stable, channel.unwrap());
    }

    #[test]
    fn calver() {
        let v = Version::CalVer((2023, 4, 0));
        let channel = Channel::try_from(&v);
        assert_eq!(Channel::Stable, channel.unwrap());
    }

    #[test]
    fn cmp_channels() {
        assert!(Channel::Stable > Channel::Rc);
        assert!(Channel::Rc > Channel::Beta);
        assert!(Channel::Beta > Channel::Alpha);
        assert!(Channel::Alpha > Channel::Dev);
    }

    #[test]
    fn cmp_parsed_versions() {
        let stable = Channel::try_from(&Version::SemVer(Semver::parse("1.1.0").unwrap())).unwrap();
        let rc = Channel::try_from(&Version::SemVer(Semver::parse("1.1.0-rc01").unwrap())).unwrap();
        let alpha =
            Channel::try_from(&Version::SemVer(Semver::parse("1.1.0-alpha01").unwrap())).unwrap();
        let beta =
            Channel::try_from(&Version::SemVer(Semver::parse("1.1.0-beta01").unwrap())).unwrap();
        assert!(Channel::Stable >= stable);
        assert!(Channel::Stable >= rc);
        assert!(Channel::Stable >= alpha);
        assert!(Channel::Stable >= beta);
        assert!(Channel::Rc >= rc);
        assert!(Channel::Rc >= beta);
        assert!(Channel::Rc >= alpha);
        assert!(Channel::Beta >= beta);
        assert!(Channel::Beta >= alpha);
        assert!(Channel::Alpha >= alpha);
    }
}
