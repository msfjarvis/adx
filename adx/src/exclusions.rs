use crate::parse::get_packages;
use clap::{ValueEnum, builder::PossibleValue};
use std::fmt::Write;

#[derive(Debug, Eq, PartialEq, PartialOrd, Clone, Copy)]
pub(crate) enum PrintType {
    IncludeGroup,
    IncludeModule,
}

impl ValueEnum for PrintType {
    fn value_variants<'a>() -> &'a [Self] {
        &[Self::IncludeGroup, Self::IncludeModule]
    }

    fn to_possible_value(&self) -> Option<clap::builder::PossibleValue> {
        match self {
            PrintType::IncludeGroup => Some(PossibleValue::new("group")),
            PrintType::IncludeModule => Some(PossibleValue::new("module")),
        }
    }
}

pub(crate) async fn print_inclusions(print_type: PrintType) {
    let packages = get_packages().await;
    let Ok(packages) = packages else { return };
    let mut rules = String::new();
    match print_type {
        PrintType::IncludeGroup => {
            packages
                .into_iter()
                .map(|pkg| pkg.group_id)
                .collect::<std::collections::HashSet<_>>()
                .into_iter()
                .for_each(|group_id| {
                    let _ = writeln!(rules, "includeGroup(\"{group_id}\")");
                });
        }
        PrintType::IncludeModule => {
            for pkg in packages {
                let _ = writeln!(
                    rules,
                    "includeModule(\"{}\", \"{}\")",
                    pkg.group_id, pkg.artifact_id
                );
            }
        }
    }
    println!(
        "
    dependencyResolutionManagement {{
      repositories {{
        exclusiveContent {{
          forRepository {{ google() }}
          filter {{
            {rules}
          }}
        }}
      }}
    }}
    ",
    );
}
