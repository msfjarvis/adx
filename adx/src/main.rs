mod channel;
mod exclusions;
mod package;
mod parse;
mod version;

use crate::exclusions::print_inclusions;
use channel::Channel;
use clap::Parser;
use color_eyre::Result;
use exclusions::PrintType;

#[derive(Parser)]
#[command(author, version, about)]
pub(crate) struct Cli {
    /// search term to filter packages with
    #[arg(required = true)]
    pub(crate) search_term: String,
    /// the release channel to find packages from
    #[arg(value_enum, long, short, default_value_t = Channel::Alpha)]
    pub(crate) channel: Channel,
    /// search term to filter packages with
    #[arg(required = false, long, default_value_t = false)]
    pub(crate) print_includes: bool,
    /// stuff
    #[arg(value_enum, long, default_value_t = PrintType::IncludeGroup)]
    pub(crate) print_type: PrintType,
}

#[tokio::main]
async fn main() -> Result<()> {
    color_eyre::install()?;
    let cli = Cli::parse();
    if cli.print_includes {
        print_inclusions(cli.print_type).await;
        return Ok(());
    }
    let packages = parse::parse(&cli.search_term, cli.channel).await?;
    if packages.is_empty() {
        println!("No results found!");
    } else {
        for package in &packages {
            println!("{package}");
        }
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use super::Cli;
    use insta_cmd::{assert_cmd_snapshot, get_cargo_bin};
    use std::process::Command;

    #[test]
    fn cli_assert() {
        <Cli as clap::CommandFactory>::command().debug_assert();
    }

    #[test]
    fn cli_help() {
        assert_cmd_snapshot!(Command::new(get_cargo_bin("adx")).arg("--help"));
    }

    #[test]
    fn cli_search() {
        assert_cmd_snapshot!(Command::new(get_cargo_bin("adx")).arg("appcompat"));
    }

    #[test]
    fn cli_search_stable() {
        assert_cmd_snapshot!(
            Command::new(get_cargo_bin("adx"))
                .arg("--channel")
                .arg("stable")
                .arg("appcompat")
        );
    }

    #[test]
    fn cli_search_no_results() {
        assert_cmd_snapshot!(Command::new(get_cargo_bin("adx")).arg("qtc"));
    }
}
