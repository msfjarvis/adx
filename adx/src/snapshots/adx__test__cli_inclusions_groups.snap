---
source: adx/src/main.rs
info:
  program: adx
  args:
    - "''"
    - "--print-includes"
    - "--print-type"
    - group
---
success: true
exit_code: 0
----- stdout -----
includeGroup("android.arch.core")
includeGroup("android.arch.lifecycle")
includeGroup("android.arch.navigation")
includeGroup("android.arch.paging")
includeGroup("android.arch.persistence")
includeGroup("android.arch.persistence.room")
includeGroup("android.arch.work")
includeGroup("androidx.activity")
includeGroup("androidx.ads")
includeGroup("androidx.annotation")
includeGroup("androidx.appcompat")
includeGroup("androidx.appfunctions")
includeGroup("androidx.appsearch")
includeGroup("androidx.arch.core")
includeGroup("androidx.asynclayoutinflater")
includeGroup("androidx.autofill")
includeGroup("androidx.baselineprofile")
includeGroup("androidx.baselineprofile.apptarget")
includeGroup("androidx.baselineprofile.consumer")
includeGroup("androidx.baselineprofile.producer")
includeGroup("androidx.benchmark")
includeGroup("androidx.biometric")
includeGroup("androidx.bluetooth")
includeGroup("androidx.browser")
includeGroup("androidx.camera")
includeGroup("androidx.camera.featurecombinationquery")
includeGroup("androidx.camera.media3")
includeGroup("androidx.camera.viewfinder")
includeGroup("androidx.car")
includeGroup("androidx.car.app")
includeGroup("androidx.cardview")
includeGroup("androidx.collection")
includeGroup("androidx.compose")
includeGroup("androidx.compose.animation")
includeGroup("androidx.compose.compiler")
includeGroup("androidx.compose.foundation")
includeGroup("androidx.compose.material")
includeGroup("androidx.compose.material3")
includeGroup("androidx.compose.material3.adaptive")
includeGroup("androidx.compose.runtime")
includeGroup("androidx.compose.ui")
includeGroup("androidx.concurrent")
includeGroup("androidx.constraintlayout")
includeGroup("androidx.contentpager")
includeGroup("androidx.coordinatorlayout")
includeGroup("androidx.core")
includeGroup("androidx.core.uwb")
includeGroup("androidx.credentials")
includeGroup("androidx.credentials.providerevents")
includeGroup("androidx.credentials.registry")
includeGroup("androidx.cursoradapter")
includeGroup("androidx.customview")
includeGroup("androidx.databinding")
includeGroup("androidx.datastore")
includeGroup("androidx.documentfile")
includeGroup("androidx.draganddrop")
includeGroup("androidx.drawerlayout")
includeGroup("androidx.dynamicanimation")
includeGroup("androidx.emoji")
includeGroup("androidx.emoji2")
includeGroup("androidx.enterprise")
includeGroup("androidx.exifinterface")
includeGroup("androidx.fragment")
includeGroup("androidx.games")
includeGroup("androidx.gaming")
includeGroup("androidx.glance")
includeGroup("androidx.gradle")
includeGroup("androidx.graphics")
includeGroup("androidx.gridlayout")
includeGroup("androidx.health")
includeGroup("androidx.health.connect")
includeGroup("androidx.heifwriter")
includeGroup("androidx.hilt")
includeGroup("androidx.ink")
includeGroup("androidx.input")
includeGroup("androidx.interpolator")
includeGroup("androidx.javascriptengine")
includeGroup("androidx.leanback")
includeGroup("androidx.legacy")
includeGroup("androidx.lifecycle")
includeGroup("androidx.lint")
includeGroup("androidx.loader")
includeGroup("androidx.localbroadcastmanager")
includeGroup("androidx.media")
includeGroup("androidx.media2")
includeGroup("androidx.media3")
includeGroup("androidx.mediarouter")
includeGroup("androidx.metrics")
includeGroup("androidx.multidex")
includeGroup("androidx.navigation")
includeGroup("androidx.navigation.safeargs")
includeGroup("androidx.navigation.safeargs.kotlin")
includeGroup("androidx.navigation3")
includeGroup("androidx.navigationevent")
includeGroup("androidx.paging")
includeGroup("androidx.palette")
includeGroup("androidx.pdf")
includeGroup("androidx.percentlayout")
includeGroup("androidx.performance")
includeGroup("androidx.photopicker")
includeGroup("androidx.preference")
includeGroup("androidx.print")
includeGroup("androidx.privacysandbox.activity")
includeGroup("androidx.privacysandbox.ads")
includeGroup("androidx.privacysandbox.library")
includeGroup("androidx.privacysandbox.plugins")
includeGroup("androidx.privacysandbox.sdkruntime")
includeGroup("androidx.privacysandbox.tools")
includeGroup("androidx.privacysandbox.ui")
includeGroup("androidx.profileinstaller")
includeGroup("androidx.recommendation")
includeGroup("androidx.recyclerview")
includeGroup("androidx.remotecallback")
includeGroup("androidx.resourceinspection")
includeGroup("androidx.room")
includeGroup("androidx.savedstate")
includeGroup("androidx.security")
includeGroup("androidx.sharetarget")
includeGroup("androidx.slice")
includeGroup("androidx.slidingpanelayout")
includeGroup("androidx.sqlite")
includeGroup("androidx.startup")
includeGroup("androidx.swiperefreshlayout")
includeGroup("androidx.test")
includeGroup("androidx.test.espresso")
includeGroup("androidx.test.espresso.idling")
includeGroup("androidx.test.ext")
includeGroup("androidx.test.janktesthelper")
includeGroup("androidx.test.services")
includeGroup("androidx.test.uiautomator")
includeGroup("androidx.textclassifier")
includeGroup("androidx.tracing")
includeGroup("androidx.transition")
includeGroup("androidx.tv")
includeGroup("androidx.tvprovider")
includeGroup("androidx.ui")
includeGroup("androidx.vectordrawable")
includeGroup("androidx.versionedparcelable")
includeGroup("androidx.viewpager")
includeGroup("androidx.viewpager2")
includeGroup("androidx.wear")
includeGroup("androidx.wear.compose")
includeGroup("androidx.wear.protolayout")
includeGroup("androidx.wear.tiles")
includeGroup("androidx.wear.watchface")
includeGroup("androidx.wear.watchfacepush")
includeGroup("androidx.webkit")
includeGroup("androidx.window")
includeGroup("androidx.window.extensions.core")
includeGroup("androidx.work")
includeGroup("androidx.xr.arcore")
includeGroup("androidx.xr.compose")
includeGroup("androidx.xr.compose.material3")
includeGroup("androidx.xr.runtime")
includeGroup("androidx.xr.scenecore")
includeGroup("com.android")
includeGroup("com.android.ai-pack")
includeGroup("com.android.application")
includeGroup("com.android.art")
includeGroup("com.android.asset-pack")
includeGroup("com.android.asset-pack-bundle")
includeGroup("com.android.billingclient")
includeGroup("com.android.car.setupwizardlib")
includeGroup("com.android.car.ui")
includeGroup("com.android.compose.screenshot")
includeGroup("com.android.databinding")
includeGroup("com.android.designcompose")
includeGroup("com.android.dynamic-feature")
includeGroup("com.android.experimental.built-in-kotlin")
includeGroup("com.android.extensions.xr")
includeGroup("com.android.fused-library")
includeGroup("com.android.identity")
includeGroup("com.android.installreferrer")
includeGroup("com.android.internal.settings")
includeGroup("com.android.java.tools.build")
includeGroup("com.android.kotlin.multiplatform.library")
includeGroup("com.android.legacy-kapt")
includeGroup("com.android.library")
includeGroup("com.android.lint")
includeGroup("com.android.ndk.thirdparty")
includeGroup("com.android.privacy-sandbox-sdk")
includeGroup("com.android.reporting")
includeGroup("com.android.security.autorepro")
includeGroup("com.android.security.autorepro.apptest")
includeGroup("com.android.security.autorepro.javahosttest")
includeGroup("com.android.security.autorepro.ndktest")
includeGroup("com.android.security.autorepro.submission")
includeGroup("com.android.security.lint")
includeGroup("com.android.settings")
includeGroup("com.android.support")
includeGroup("com.android.support.constraint")
includeGroup("com.android.support.test")
includeGroup("com.android.support.test.espresso")
includeGroup("com.android.support.test.espresso.idling")
includeGroup("com.android.support.test.janktesthelper")
includeGroup("com.android.support.test.services")
includeGroup("com.android.support.test.uiautomator")
includeGroup("com.android.test")
includeGroup("com.android.tools")
includeGroup("com.android.tools.adblib")
includeGroup("com.android.tools.analytics-library")
includeGroup("com.android.tools.apkdeployer")
includeGroup("com.android.tools.apkparser")
includeGroup("com.android.tools.build")
includeGroup("com.android.tools.build.jetifier")
includeGroup("com.android.tools.chunkio")
includeGroup("com.android.tools.compose")
includeGroup("com.android.tools.ddms")
includeGroup("com.android.tools.emulator")
includeGroup("com.android.tools.external.com-intellij")
includeGroup("com.android.tools.external.org-jetbrains")
includeGroup("com.android.tools.fakeadbserver")
includeGroup("com.android.tools.internal.build.test")
includeGroup("com.android.tools.journeys")
includeGroup("com.android.tools.layoutlib")
includeGroup("com.android.tools.lint")
includeGroup("com.android.tools.metalava")
includeGroup("com.android.tools.pixelprobe")
includeGroup("com.android.tools.screenshot")
includeGroup("com.android.tools.smali")
includeGroup("com.android.tools.utp")
includeGroup("com.android.volley")
includeGroup("com.crashlytics.sdk.android")
includeGroup("com.google.ads.afsn")
includeGroup("com.google.ads.interactivemedia.v3")
includeGroup("com.google.ads.mediation")
includeGroup("com.google.ai.client.generativeai")
includeGroup("com.google.ai.edge.aicore")
includeGroup("com.google.ai.edge.litert")
includeGroup("com.google.ai.edge.localagents")
includeGroup("com.google.ambient.crossdevice")
includeGroup("com.google.android.ads")
includeGroup("com.google.android.ads.consent")
includeGroup("com.google.android.appcrawler")
includeGroup("com.google.android.apps.common.testing.accessibility.framework")
includeGroup("com.google.android.car.connectionservice")
includeGroup("com.google.android.datatransport")
includeGroup("com.google.android.engage")
includeGroup("com.google.android.enterprise.connectedapps")
includeGroup("com.google.android.exoplayer")
includeGroup("com.google.android.fhir")
includeGroup("com.google.android.flexbox")
includeGroup("com.google.android.games")
includeGroup("com.google.android.gms")
includeGroup("com.google.android.gms.oss-licenses-plugin")
includeGroup("com.google.android.gms.strict-version-matcher-plugin")
includeGroup("com.google.android.instantapps")
includeGroup("com.google.android.instantapps.thirdpartycompat")
includeGroup("com.google.android.libraries.ads.mobile.sdk")
includeGroup("com.google.android.libraries.car")
includeGroup("com.google.android.libraries.cloud.telco.subgraph")
includeGroup("com.google.android.libraries.enterprise.amapi")
includeGroup("com.google.android.libraries.healthdata")
includeGroup("com.google.android.libraries.identity.googleid")
includeGroup("com.google.android.libraries.maps")
includeGroup("com.google.android.libraries.mapsplatform.secrets-gradle-plugin")
includeGroup("com.google.android.libraries.mapsplatform.transportation")
includeGroup("com.google.android.libraries.navigation")
includeGroup("com.google.android.libraries.places")
includeGroup("com.google.android.libraries.play.games")
includeGroup("com.google.android.libraries.sdkcoroutines")
includeGroup("com.google.android.libraries.searchinapps")
includeGroup("com.google.android.livesharing")
includeGroup("com.google.android.material")
includeGroup("com.google.android.mdoc")
includeGroup("com.google.android.mediahome")
includeGroup("com.google.android.meet")
includeGroup("com.google.android.odml")
includeGroup("com.google.android.play")
includeGroup("com.google.android.recaptcha")
includeGroup("com.google.android.support")
includeGroup("com.google.android.things")
includeGroup("com.google.android.tv")
includeGroup("com.google.android.tv.tvpanelframework")
includeGroup("com.google.android.ump")
includeGroup("com.google.android.wearable")
includeGroup("com.google.android.wearable.watchface.validator")
includeGroup("com.google.androidbrowserhelper")
includeGroup("com.google.ar")
includeGroup("com.google.ar.sceneform")
includeGroup("com.google.ar.sceneform.ux")
includeGroup("com.google.assistant.appactions")
includeGroup("com.google.assistant.suggestion")
includeGroup("com.google.camerax.effects")
includeGroup("com.google.chromeos")
includeGroup("com.google.cose")
includeGroup("com.google.d2c")
includeGroup("com.google.devtools.ksp")
includeGroup("com.google.fhir")
includeGroup("com.google.firebase")
includeGroup("com.google.firebase.appdistribution")
includeGroup("com.google.firebase.crashlytics")
includeGroup("com.google.firebase.firebase-perf")
includeGroup("com.google.firebase.testlab")
includeGroup("com.google.gms")
includeGroup("com.google.gms.google-services")
includeGroup("com.google.jacquard")
includeGroup("com.google.mediapipe")
includeGroup("com.google.mlkit")
includeGroup("com.google.net.cronet")
includeGroup("com.google.oboe")
includeGroup("com.google.play.policy.insights")
includeGroup("com.google.prefab")
includeGroup("com.google.relay")
includeGroup("com.google.test.platform")
includeGroup("com.google.testing.platform")
includeGroup("io.fabric.sdk.android")
includeGroup("org.chromium.net")
includeGroup("org.multipaz")
includeGroup("tools.base.build-system.debug")
includeGroup("zipflinger")


----- stderr -----
